### PR TITLE
Automatic vocabulary size padding

### LINKED ISSUE #221: Suggest multiple of 64 for vocab size during training and finetuning
https://twitter.com/karpathy/status/1621578354024677377

### PR DESCRIPTION
**Pull request by carmocca**
Fixes #221

Part of #222

### CODE REVIEW

File: lit_llama/model.py [hunk 1]
```diff
@@ -35,14 +44,13 @@ def from_name(cls, name: str) -> Self:
...
-        assert config.block_size is not None
+        assert config.padded_vocab_size is not None
         self.config = config
 
-        self.lm_head = nn.Linear(config.n_embd, config.vocab_size, bias=False)
+        self.lm_head = nn.Linear(config.n_embd, config.padded_vocab_size, bias=False)
         self.transformer = nn.ModuleDict(
             dict(
-                wte=nn.Embedding(config.vocab_size, config.n_embd),
+                wte=nn.Embedding(config.padded_vocab_size, config.n_embd),
```
Comment (awaelchli) [inline lit_llama/model.py:53]: The current default vocab size is a multiple of 64, so this should be save since the tensor size here won't actually be different to the shape in the checkpoint.

File: lit_llama/model.py [hunk 2]
```diff
@@ -103,7 +111,7 @@ def __init__(self, config: LLaMAConfig) -> None:
         self.n_head = config.n_head
         self.n_embd = config.n_embd
         self.block_size = config.block_size
-        self.rope_cache = None
+        self.rope_cache: Optional[torch.Tensor] = None
```
Comment (awaelchli) [inline lit_llama/model.py:114]: I wouldn't add the typing 😅

**Review by awaelchli — APPROVED** [review]
